### PR TITLE
Angrimson bug fix collection

### LINKED ISSUE #1306: [bug] Some generated configs don't satisfy forbiddens
I think this happens because when sampling the initial configs from a `RandomInitialDesign`, `configspace.create_uniform_configspace_copy` is called to instantiate a modified copy of the search space (which replaces each hyperparameter with uniform versions thereof), while conditions are [explicitly set](https://github.com/automl/SMAC3/blob/b8f10493df802def001ec32c666ed03533f3bb06/smac/utils/configspace.py#L280) on the newly created space, setting forbiddens is omitted.

### LINKED ISSUE #1336: [BUG] SMAC3 not compatible with sklearn 1.9 due to changes in their tree models (RF)
There have been changes to the tree  models in sklearn 1.9, making it incompatible with smac3 ("cannot import name 'DTYPE' from 'sklearn.tree._tree')

#### Description
I tried to run smac3 with sklearn 1.9

#### Steps/Code to Reproduce

from ConfigSpace import Configuration, ConfigurationSpace

import numpy as np
from smac import HyperparameterOptimizationFacade, Scenario
from sklearn import datasets
from sklearn.svm import SVC
from sklearn.model_selection import cross_val_score

iris = datasets.load_iris()

def train(config: Configuration, seed: int = 0) -> float:
    classifier = SVC(C=config["C"], random_state=seed)
    scores = cross_val_score(classifier, iris.data, iris.target, cv=5)
    return 1 - np.mean(scores)

configspace = ConfigurationSpace({"C": (0.100, 1000.0)})

scenario = Scenario(configspace, deterministic=True, n_trials=200)

smac = HyperparameterOptimizationFacade(scenario, train)
incumbent = smac.optimize()

#### Expected Results
No error message

#### Actual Results
{
	"name": "ImportError",
	"message": "cannot import name 'DTYPE' from 'sklearn.tree._tree' (/home/sagemaker-user/MotherML/.venv/lib/python3.12/site-packages/sklearn/tree/_tree.cpython-312-x86_64-linux-gnu.so)",
	"stack": "\u001b[31m---------------------------------------------------------------------------

#### Versions
smac==2.4.0

### PR DESCRIPTION
**Pull request by angrinord**
2 bug fixes cherry-picked from dev branch.

#### Type Of Changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

- [ ] My change is based on the latest stage of the `develop` branch.
- [ ] My change required a change of the documentation, which has been done.
- [ ] I checked that the documentation can be build, visualizes everything as expected, and does not contain any warnings.
- [ ] I have added/adapted tests to cover my changes.
- [ ] The tests can be executed successfully.
- [ ] I have added a description of the changes to `CHANGELOG.rst`.

#### Any other comments?
I suggest updating version number to v2.4.1

### CODE REVIEW

**Comment by angrinord** [comment]
This should also stop main from failing its scheduled tests.

File: smac/utils/configspace.py [hunk 1]
```diff
@@ -278,4 +278,9 @@ def create_uniform_configspace_copy(
         new_configuration_space.add(new_hyperparameter)
     conditions = configspace.conditions
     new_configuration_space.add(conditions)
+    forbiddens = configspace.forbidden_clauses
```
Comment (kalifou) [inline smac/utils/configspace.py:285]: Hi!
Do I understand correctly that, generally speaking, the function is creating a copy for the HP config space with uniform HP distribution, and here in particular, you want to also copy its forbidden HP clauses (as long as they exist ) using this particular substitute_hyperparameters_in_forbiddens function? 

IMO, one could benefit from a couple of lines of comment, if possible :-) Maybe just in its docstring or somewhere in between, to describe what this addition does?

**Comment by kalifou** [comment]
Hi! 
The PR generally looks good to me. As mentioned above, since the adds are technical and specific, I thought a couple of lines of comments could help clarify the fix.
Thanks

**Review by mwever — APPROVED** [review]
LG2M

**Review by mwever — CHANGES_REQUESTED** [review]
Can you please add some logs to the CHANGELOG.md for this bugfix?

File: smac/utils/configspace.py (hunk 1 above)
Comment (angrinord) [reply smac/utils/configspace.py:285]: I added it to the docstring

**Review by mwever — APPROVED** [review]

**Review by kalifou — APPROVED** [review]
LG2M
